File: ksit_nexus_app/lib/services/quiet_hours_service.dart
/// Quiet hours service for managing notification quiet hours
import 'package:shared_preferences/shared_preferences.dart';
import '../utils/logger.dart';

final appLogger = Logger('QuietHoursService');

class QuietHoursService {
  static final QuietHoursService _instance = QuietHoursService._internal();
  factory QuietHoursService() => _instance;
  QuietHoursService._internal();

  SharedPreferences? _prefs;
  static const String _enabledKey = 'quiet_hours_enabled';
  static const String _startHourKey = 'quiet_hours_start_hour';
  static const String _startMinuteKey = 'quiet_hours_start_minute';
  static const String _endHourKey = 'quiet_hours_end_hour';
  static const String _endMinuteKey = 'quiet_hours_end_minute';

  Future<void> init() async {
    _prefs = await SharedPreferences.getInstance();
  }

  /// Check if quiet hours are enabled
  Future<bool> isEnabled() async {
    try {
      return _prefs?.getBool(_enabledKey) ?? false;
    } catch (e) {
      appLogger.error('Error checking quiet hours enabled: $e');
      return false;
    }
  }

  /// Enable quiet hours
  Future<bool> enable() async {
    try {
      return await _prefs?.setBool(_enabledKey, true) ?? false;
    } catch (e) {
      appLogger.error('Error enabling quiet hours: $e');
      return false;
    }
  }

  /// Disable quiet hours
  Future<bool> disable() async {
    try {
      return await _prefs?.setBool(_enabledKey, false) ?? false;
    } catch (e) {
      appLogger.error('Error disabling quiet hours: $e');
      return false;
    }
  }

  /// Set quiet hours
  Future<bool> setQuietHours({
    required int startHour,
    required int startMinute,
    required int endHour,
    required int endMinute,
  }) async {
    try {
      await _prefs?.setInt(_startHourKey, startHour);
      await _prefs?.setInt(_startMinuteKey, startMinute);
      await _prefs?.setInt(_endHourKey, endHour);
      await _prefs?.setInt(_endMinuteKey, endMinute);
      return true;
    } catch (e) {
      appLogger.error('Error setting quiet hours: $e');
      return false;
    }
  }

  /// Get quiet hours
  Future<Map<String, int>?> getQuietHours() async {
    try {
      final startHour = _prefs?.getInt(_startHourKey);
      final startMinute = _prefs?.getInt(_startMinuteKey);
      final endHour = _prefs?.getInt(_endHourKey);
      final endMinute = _prefs?.getInt(_endMinuteKey);

      if (startHour == null ||
          startMinute == null ||
          endHour == null ||
          endMinute == null) {
        return null;
      }

      return {
        'startHour': startHour,
        'startMinute': startMinute,
        'endHour': endHour,
        'endMinute': endMinute,
      };
    } catch (e) {
      appLogger.error('Error getting quiet hours: $e');
      return null;
    }
  }

  /// Check if current time is within quiet hours
  Future<bool> isQuietHours() async {
    try {
      final enabled = await isEnabled();
      if (!enabled) {
        return false;
      }

      final hours = await getQuietHours();
      if (hours == null) {
        return false;
      }

      final now = DateTime.now();
      final currentTime = DateTime(
        now.year,
        now.month,
        now.day,
        now.hour,
        now.minute,
      );

      final startTime = DateTime(
        now.year,
        now.month,
        now.day,
        hours['startHour']!,
        hours['startMinute']!,
      );

      final endTime = DateTime(
        now.year,
        now.month,
        now.day,
        hours['endHour']!,
        hours['endMinute']!,
      );

      // Handle quiet hours that span midnight
      if (startTime.isBefore(endTime) || startTime.isAtSameMomentAs(endTime)) {
        // Normal case: quiet hours within same day
        return currentTime.isAfter(startTime) && currentTime.isBefore(endTime);
      } else {
        // Quiet hours span midnight
        return currentTime.isAfter(startTime) || currentTime.isBefore(endTime);
      }
    } catch (e) {
      appLogger.error('Error checking quiet hours: $e');
      return false;
    }
  }

  /// Check if notification should be sent based on quiet hours
  Future<bool> shouldSendNotification(String priority) async {
    // Urgent and high priority notifications bypass quiet hours
    if (priority == 'urgent' || priority == 'high') {
      return true;
    }

    // Check quiet hours for medium and low priority
    final inQuietHours = await isQuietHours();
    return !inQuietHours;
  }

  /// Get next time when notifications can be sent
  Future<DateTime?> getNextSendTime() async {
    try {
      final enabled = await isEnabled();
      if (!enabled) {
        return null;
      }

      final inQuietHours = await isQuietHours();
      if (!inQuietHours) {
        return null;
      }

      final hours = await getQuietHours();
      if (hours == null) {
        return null;
      }

      final now = DateTime.now();
      final endTime = DateTime(
        now.year,
        now.month,
        now.day,
        hours['endHour']!,
        hours['endMinute']!,
      );

      final startTime = DateTime(
        now.year,
        now.month,
        now.day,
        hours['startHour']!,
        hours['startMinute']!,
      );

      // If end time is before start time, quiet hours span midnight
      if (endTime.isBefore(startTime)) {
        // End time is next day
        if (now.isAfter(startTime)) {
          // After start, end is next day
          return DateTime(
            now.year,
            now.month,
            now.day + 1,
            hours['endHour']!,
            hours['endMinute']!,
          );
        } else {
          // Before start, end is today
          return endTime;
        }
      } else {
        // Same day
        return endTime;
      }
    } catch (e) {
      appLogger.error('Error getting next send time: $e');
      return null;
    }
  }
}

/// Global quiet hours service instance
final quietHoursService = QuietHoursService();

File: backend/apps/notifications/services/quiet_hours_service.py
"""
Quiet hours service for notification scheduling
"""
from datetime import time, datetime
from typing import Optional
from django.utils import timezone
import pytz
from ..models import NotificationPreference


class QuietHoursService:
    """Service for managing quiet hours"""
    
    @staticmethod
    def is_quiet_hours(user) -> bool:
        """
        Check if current time is within quiet hours for user
        
        Args:
            user: User instance
            
        Returns:
            True if current time is within quiet hours, False otherwise
        """
        try:
            pref = NotificationPreference.objects.filter(user=user).first()
            if not pref:
                return False
            
            # Check if quiet hours are enabled
            if not pref.quiet_hours_start or not pref.quiet_hours_end:
                return False
            
            # Get user's timezone
            if pref.timezone:
                try:
                    user_tz = pytz.timezone(pref.timezone)
                except pytz.UnknownTimeZoneError:
                    user_tz = timezone.get_current_timezone()
            else:
                user_tz = timezone.get_current_timezone()
            current_time = timezone.now().astimezone(user_tz).time()
            
            start_time = pref.quiet_hours_start
            end_time = pref.quiet_hours_end
            
            # Handle quiet hours that span midnight (e.g., 22:00 - 06:00)
            if start_time <= end_time:
                # Normal case: quiet hours within same day
                return start_time <= current_time <= end_time
            else:
                # Quiet hours span midnight
                return current_time >= start_time or current_time <= end_time
                
        except Exception as e:
            print(f"Error checking quiet hours: {e}")
            return False
    
    @staticmethod
    def should_send_notification(user, priority: str = 'medium') -> bool:
        """
        Check if notification should be sent based on quiet hours
        
        Args:
            user: User instance
            priority: Notification priority (low, medium, high, urgent)
            
        Returns:
            True if notification should be sent, False if in quiet hours
        """
        # Urgent notifications bypass quiet hours
        if priority == 'urgent':
            return True
        
        # High priority notifications bypass quiet hours
        if priority == 'high':
            return True
        
        # Check quiet hours for medium and low priority
        return not QuietHoursService.is_quiet_hours(user)
    
    @staticmethod
    def get_next_send_time(user) -> Optional[datetime]:
        """
        Get the next time when notifications can be sent (after quiet hours)
        
        Args:
            user: User instance
            
        Returns:
            Next datetime when notifications can be sent, None if quiet hours not enabled
        """
        try:
            pref = NotificationPreference.objects.filter(user=user).first()
            if not pref or not pref.quiet_hours_start or not pref.quiet_hours_end:
                return None
            
            # Get user's timezone
            if pref.timezone:
                try:
                    user_tz = pytz.timezone(pref.timezone)
                except pytz.UnknownTimeZoneError:
                    user_tz = timezone.get_current_timezone()
            else:
                user_tz = timezone.get_current_timezone()
            now = timezone.now().astimezone(user_tz)
            current_time = now.time()
            
            start_time = pref.quiet_hours_start
            end_time = pref.quiet_hours_end
            
            # If currently in quiet hours, return end time today
            if QuietHoursService.is_quiet_hours(user):
                # Calculate end time
                if start_time <= end_time:
                    # Same day
                    next_send = now.replace(
                        hour=end_time.hour,
                        minute=end_time.minute,
                        second=end_time.second,
                        microsecond=0
                    )
                else:
                    # Spans midnight - end is next day
                    if current_time >= start_time:
                        # After start, end is next day
                        next_send = (now + timezone.timedelta(days=1)).replace(
                            hour=end_time.hour,
                            minute=end_time.minute,
                            second=end_time.second,
                            microsecond=0
                        )
                    else:
                        # Before start, end is today
                        next_send = now.replace(
                            hour=end_time.hour,
                            minute=end_time.minute,
                            second=end_time.second,
                            microsecond=0
                        )
                
                return next_send.astimezone(timezone.utc)
            
            # Not in quiet hours
            return None
            
        except Exception as e:
            print(f"Error getting next send time: {e}")
            return None
    
    @staticmethod
    def set_quiet_hours(user, start_time: time, end_time: time, timezone_str: str = 'Asia/Kolkata'):
        """
        Set quiet hours for user
        
        Args:
            user: User instance
            start_time: Start time for quiet hours
            end_time: End time for quiet hours
            timezone_str: Timezone string
            
        Returns:
            NotificationPreference instance
        """
        pref, created = NotificationPreference.objects.get_or_create(
            user=user,
            defaults={
                'quiet_hours_start': start_time,
                'quiet_hours_end': end_time,
                'timezone': timezone_str,
            }
        )
        
        if not created:
            pref.quiet_hours_start = start_time
            pref.quiet_hours_end = end_time
            pref.timezone = timezone_str
            pref.save()
        
        return pref
    
    @staticmethod
    def disable_quiet_hours(user):
        """
        Disable quiet hours for user
        
        Args:
            user: User instance
        """
        pref = NotificationPreference.objects.filter(user=user).first()
        if pref:
            pref.quiet_hours_start = None
            pref.quiet_hours_end = None
            pref.save()

